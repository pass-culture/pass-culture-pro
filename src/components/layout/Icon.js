import PropTypes from 'prop-types'
import React from 'react'

import { ROOT_PATH } from '../../utils/config'

export const Icon = ({ svg, ...imgProps }) => (
  <img alt={svg} src={`${ROOT_PATH}/icons/${svg}.svg`} {...imgProps} />
)

Icon.propTypes = {
  svg: PropTypes.string.isRequired,
}

export default Icon
