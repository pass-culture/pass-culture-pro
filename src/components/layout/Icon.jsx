import PropTypes from 'prop-types'
import React from 'react'

import { ROOT_PATH } from '../../utils/config'

export const getImageUrl = (svg, png) => {
  let iconUrl
  svg ? (iconUrl = `${ROOT_PATH}/icons/${svg}.svg`) : (iconUrl = `${ROOT_PATH}/icons/${png}.png`)
  return iconUrl
}

const Icon = ({ png, svg, ...imgProps }) => {
  let altMessage
  const iconUrl = getImageUrl(svg, png)
  svg ? (altMessage = svg) : (altMessage = png)
  return (<img
    alt={altMessage}
    src={iconUrl}
    {...imgProps}
          />)
}

Icon.defaultProps = {
  png: null,
}

Icon.propTypes = {
  png: PropTypes.string,
  svg: PropTypes.string.isRequired,
}

export default Icon
