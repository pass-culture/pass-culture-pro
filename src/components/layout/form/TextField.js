/* eslint
  react/jsx-one-expression-per-line: 0 */
import classnames from 'classnames'
import React from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'
import {
  composeValidators,
  config,
  createParseNumberValue,
  createValidateRequiredField,
} from 'react-final-form-utils'

import FieldError from './FieldError'

const validateRequiredField = createValidateRequiredField(
  config.DEFAULT_REQUIRED_ERROR
)

export const TextField = ({
  autoComplete,
  className,
  disabled,
  format,
  id,
  label,
  name,
  parse,
  placeholder,
  readOnly,
  renderInner,
  renderValue,
  required,
  title,
  type,
  validate,
  ...inputProps
}) => {
  const requiredValidate =
    required && typeof required === 'function'
      ? required
      : (required && validateRequiredField) || undefined

  return (
    <Field
      format={format}
      name={name}
      validate={composeValidators(validate, requiredValidate)}
      parse={parse || createParseNumberValue(type)}
      render={({ input, meta }) => {
        let value = input.value
        if (!readOnly && typeof value === 'string') {
          value = ''
        }
        return (
          <div
            className={classnames('field text-field', className, {
              'is-read-only': readOnly,
            })}
            id={id}>
            <label
              htmlFor={name}
              className={classnames('field-label', { empty: !label })}>
              {label && (
                <span>
                  <span>{label}</span>
                  {required && !readOnly && (
                    <span className="field-asterisk">*</span>
                  )}
                </span>
              )}
            </label>
            <div className="field-control">
              <div className="field-value flex-columns items-center">
                <div className="field-inner flex-columns items-center">
                  <input
                    {...input}
                    {...inputProps}
                    autoComplete={autoComplete ? 'on' : 'off'}
                    className={`field-input field-${type}`}
                    disabled={disabled || readOnly}
                    placeholder={readOnly ? '' : placeholder}
                    readOnly={readOnly}
                    required={!!required} // cast to boolean
                    title={title}
                    type={readOnly ? 'text' : type}
                    value={value}
                  />
                  {renderInner()}
                </div>
                {renderValue()}
              </div>
              <FieldError meta={meta} />
            </div>
          </div>
        )
      }}
    />
  )
}

TextField.defaultProps = {
  autoComplete: false,
  className: '',
  disabled: false,
  format: null,
  id: null,
  label: '',
  parse: null,
  placeholder: 'Please enter a value',
  readOnly: false,
  renderInner: () => null,
  renderValue: () => null,
  required: false,
  title: null,
  type: 'text',
  validate: null,
}

TextField.propTypes = {
  autoComplete: PropTypes.bool,
  className: PropTypes.string,
  disabled: PropTypes.bool,
  format: PropTypes.func,
  id: PropTypes.string,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  parse: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderInner: PropTypes.func,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  title: PropTypes.string,
  type: PropTypes.string,
  validate: PropTypes.func,
}

export default TextField
