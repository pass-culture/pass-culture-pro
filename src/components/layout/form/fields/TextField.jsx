import classnames from 'classnames'
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Field } from 'react-final-form'
import { composeValidators, createParseNumberValue } from 'react-final-form-utils'

import FieldErrors from '../FieldErrors'
import getRequiredValidate from '../utils/getRequiredValidate'

function getInputValue(inputType, value) {
  const isStringifiedNumber = inputType === 'number' && typeof value === 'string'

  if (isStringifiedNumber) {
    return ''
  }

  return value
}

class TextField extends PureComponent {
  componentDidMount() {
    const { type, isDecimal } = this.props
    if (type === 'number') {
      this.preventEnteringInvalidChars(isDecimal)
    }
  }

  componentWillUnmount() {
    if (this.keypressListener) {
      this.inputElement.removeEventListener(this.keypressListener)
    }
  }

  preventEnteringInvalidChars = (isDecimal) => {
    this.keypressListener = this.inputElement.addEventListener('keypress', event => {
      const validCharacters = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '0']

      if (isDecimal) {
        validCharacters.push(',', '.')
      }

      if (!validCharacters.includes(event.key)) {
        event.preventDefault()
      }
    })
  }

  handleRef = _e => {
    this.inputElement = _e
  }

  renderField = ({ input, meta }) => {
    const {
      className,
      disabled,
      id,
      innerClassName,
      isDecimal,
      label,
      min,
      name,
      placeholder,
      readOnly,
      renderInner,
      renderValue,
      required,
      type,
      ...inputProps
    } = this.props
    const inputType = readOnly ? 'text' : type
    const inputValue = getInputValue(inputType, input.value)

    return (
      <div
        className={classnames('field text-field', className, {
          'is-label-aligned': label,
          'is-read-only': readOnly,
        })}
        id={id}
      >
        {label && (
          <label
            className={classnames('field-label')}
            htmlFor={name}
          >
            <span>
              <span>
                {label}
              </span>
              {required && !readOnly && (
                <span className="field-asterisk">
                  {'*'}
                </span>
              )}
            </span>
          </label>
        )}
        <div className="field-control">
          <div className="field-value flex-columns items-center">
            <div className={classnames('field-inner flex-columns items-center', innerClassName)}>
              <input
                id={name}
                {...input}
                {...inputProps}
                className={`field-input field-${type}`}
                disabled={disabled || readOnly}
                min={min}
                placeholder={readOnly ? '' : placeholder}
                readOnly={readOnly}
                ref={this.handleRef}
                required={!!required}
                type={inputType}
                value={inputValue}
              />
              {renderInner()}
            </div>
            {renderValue()}
          </div>
          <FieldErrors meta={meta} />
        </div>
        <div />
      </div>
    )
  }

  render() {
    const { format, name, parse, required, type, validate } = this.props

    return (
      <Field
        format={format}
        name={name}
        parse={parse || createParseNumberValue(type)}
        render={this.renderField}
        validate={composeValidators(validate, getRequiredValidate(required, type))}
      />
    )
  }
}

TextField.defaultProps = {
  className: '',
  disabled: false,
  format: null,
  id: null,
  innerClassName: null,
  isDecimal: true,
  label: '',
  parse: null,
  placeholder: '',
  readOnly: false,
  renderInner: () => null,
  renderValue: () => null,
  required: false,
  type: 'text',
  validate: null,
}

TextField.propTypes = {
  className: PropTypes.string,
  disabled: PropTypes.bool,
  format: PropTypes.func,
  id: PropTypes.string,
  innerClassName: PropTypes.string,
  isDecimal: PropTypes.bool,
  label: PropTypes.string,
  name: PropTypes.string.isRequired,
  parse: PropTypes.func,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderInner: PropTypes.func,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  type: PropTypes.string,
  validate: PropTypes.func,
}

export default TextField
