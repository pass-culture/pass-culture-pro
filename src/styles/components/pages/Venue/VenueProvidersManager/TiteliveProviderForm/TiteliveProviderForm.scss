.titelive-provider-form {
  display: flex;
  flex-direction: row;

  .compte-section {
    flex-direction: column;
    max-width: 250px;
    margin-right: 20px;

    .label-section {
      display: flex;
      flex-direction: row;
      justify-content: space-around;
      align-items: baseline;

      .label-text {
        vertical-align: center;
        font-size: 12px;
      }
    }

  }

  .provider-import-button-container {
    position: relative;

    button {
      position: absolute;
      top: 48%;
    }
  }
}
