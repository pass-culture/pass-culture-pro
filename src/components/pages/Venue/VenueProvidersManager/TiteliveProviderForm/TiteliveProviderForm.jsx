import classNames from 'classnames'
import React, {PureComponent} from 'react'
import {Form} from 'react-final-form'

import TextField from '../../../../layout/form/fields/TextField'
import Icon from '../../../../layout/Icon'
import PropTypes from 'prop-types'

class TiteliveProviderForm extends PureComponent {
  constructor() {
    super()
    this.state = {
      isLoadingMode: false,
    }
  }

  handleFormSubmit = (formValues, form) => {
    formValues.preventDefault()
    this.setState({ isLoadingMode: true })

    const { createVenueProvider } = this.props
    const { venueIdAtOfferProvider } = formValues
    const { providerId, venueId } = this.props

    const payload = {
      providerId: providerId,
      venueIdAtOfferProvider,
      venueId: venueId,
    }

    return createVenueProvider(this.handleFail(form), this.handleSuccess, payload)
  }

  handleSuccess = () => {
    const {
      history,
      offererId,
      venueId,
    } = this.props
    history.push(`/structures/${offererId}/lieux/${venueId}`)
  }

  handleFail = () => (state, action) => {
    const { notify } = this.props
    const {
      payload: { errors },
    } = action

    notify(errors)
  }

  renderForm = props => {
    const { venueIdAtOfferProviderIsRequired } = this.props
    const { isLoadingMode } = this.state
    return (
      <form onSubmit={props.handleSubmit}>
        <div className="titelive-provider-form">
          <div className="compte-section">
            <div className="label-section">
              <label htmlFor="venueIdAtOfferProvider">
                {'Compte'}
              </label>
              {!isLoadingMode && venueIdAtOfferProviderIsRequired && (
                <span
                  className="tooltip tooltip-info"
                  data-place="bottom"
                  data-tip={`<p>Veuillez saisir un compte.</p>`}
                >
                  <Icon
                    alt="image d’aide à l’information"
                    svg="picto-info"
                  />
                </span>
              )}
            </div>

            <TextField
              className={classNames('field-text', {
                'field-is-read-only': !venueIdAtOfferProviderIsRequired || isLoadingMode,
              })}
              name="venueIdAtOfferProvider"
              readOnly={!venueIdAtOfferProviderIsRequired || isLoadingMode}
              required
            />
          </div>

          {!isLoadingMode && (
            <div className="provider-import-button-container">
              <button
                className="button is-intermediate provider-import-button"
                type="submit"
              >
                {'Importer'}
              </button>
            </div>
          )}
        </div>
      </form>
    )
  }

  render() {
    return (<Form
      onSubmit={this.handleFormSubmit}
      render={this.renderForm}
            />)
  }
}

TiteliveProviderForm.propTypes = {
  createVenueProvider: PropTypes.func.isRequired,
  history: PropTypes.shape().isRequired,
  notify: PropTypes.func.isRequired,
  providerId: PropTypes.string.isRequired,
  venueId: PropTypes.string.isRequired,
  venueIdAtOfferProviderIsRequired: PropTypes.bool.isRequired,
}

export default TiteliveProviderForm
