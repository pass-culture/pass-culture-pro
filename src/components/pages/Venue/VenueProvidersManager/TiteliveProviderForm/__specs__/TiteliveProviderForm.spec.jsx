import React from 'react'
import { mount, shallow } from 'enzyme'

import TiteliveProviderForm from '../TiteliveProviderForm'
import TextField from '../../../../../layout/form/fields/TextField'
import Icon from '../../../../../layout/Icon'
import { Form } from 'react-final-form'

describe('src | components | pages | Venue | VenueProvidersManager | form | TiteliveProviderForm', () => {
  let createVenueProvider
  let props
  let notify
  let history

  beforeEach(() => {
    createVenueProvider = jest.fn()
    history = {
      push: jest.fn(),
    }
    notify = jest.fn()
    props = {
      createVenueProvider,
      history,
      notify,
      offererId: 'CC',
      providerId: 'CC',
      venueId: 'AA',
      venueIdAtOfferProviderIsRequired: false,
    }
  })

  it('should initialize TiteliveProviderForm component with default state', () => {
    // when
    const wrapper = shallow(<TiteliveProviderForm {...props} />)

    // then
    expect(wrapper.state()).toStrictEqual({
      isLoadingMode: false,
    })
  })

  describe('when not in loading mode', () => {
    it('should display an import button', () => {
      // when
      const wrapper = mount(<TiteliveProviderForm {...props} />)

      // then
      const importButtonContainer = wrapper.find('.provider-import-button-container')
      expect(importButtonContainer).toHaveLength(1)
      const importButton = importButtonContainer.find('button')
      expect(importButton).toHaveLength(1)
      expect(importButton.prop('className')).toBe('button is-intermediate provider-import-button')
      expect(importButton.prop('type')).toBe('submit')
      expect(importButton.text()).toBe('Importer')
    })

    describe('when provider identifier is required', () => {
      it('should render a TextField component not in read only mode', () => {
        // given
        props.venueIdAtOfferProviderIsRequired = true

        // when
        const wrapper = mount(<TiteliveProviderForm {...props} />)

        // then
        const form = wrapper.find(Form)
        expect(form).toHaveLength(1)
        const label = form.find('label')
        expect(label.text()).toBe('Compte')
        const textField = form.find(TextField)
        expect(textField).toHaveLength(1)
        expect(textField.prop('className')).toBe('field-text')
        expect(textField.prop('name')).toBe('venueIdAtOfferProvider')
        expect(textField.prop('readOnly')).toBe(false)
        expect(textField.prop('required')).toBe(true)
      })

      it('should display a tooltip and an Icon component', () => {
        // given
        props.venueIdAtOfferProviderIsRequired = true

        // when
        const wrapper = mount(<TiteliveProviderForm {...props} />)

        // then
        const tooltip = wrapper.find('.tooltip-info')
        expect(tooltip).toHaveLength(1)
        expect(tooltip.prop('className')).toBe('tooltip tooltip-info')
        expect(tooltip.prop('data-place')).toBe('bottom')
        expect(tooltip.prop('data-tip')).toBe('<p>Veuillez saisir un compte.</p>')
        const icon = tooltip.find(Icon)
        expect(icon).toHaveLength(1)
        expect(icon.prop('svg')).toBe('picto-info')
        expect(icon.prop('alt')).toBe('image d’aide à l’information')
      })
    })

    describe('when provider identifier is not required', () => {
      it('should render a TextField component in read only mode', () => {
        // given
        props.venueIdAtOfferProviderIsRequired = false

        // when
        const wrapper = mount(<TiteliveProviderForm {...props} />)

        // then
        const form = wrapper.find(Form)
        expect(form).toHaveLength(1)
        const label = form.find('label')
        expect(label.text()).toBe('Compte')
        const textField = form.find(TextField)
        expect(textField).toHaveLength(1)
        expect(textField.prop('className')).toBe('field-text field-is-read-only')
        expect(textField.prop('name')).toBe('venueIdAtOfferProvider')
        expect(textField.prop('readOnly')).toBe(true)
        expect(textField.prop('required')).toBe(true)
      })

      it('should not display a tooltip and an Icon component', () => {
        // given
        props.venueIdAtOfferProviderIsRequired = false

        // when
        const wrapper = mount(<TiteliveProviderForm {...props} />)

        // then
        const tooltip = wrapper.find('.tooltip-info')
        expect(tooltip).toHaveLength(0)
      })
    })
  })

  describe('handleSubmit', () => {
    it('should update venue provider using API', () => {
      // given
      const formValues = {
        venueIdAtOfferProvider: 'token',
        preventDefault: jest.fn(),
      }
      const wrapper = shallow(<TiteliveProviderForm {...props} />)

      // when
      wrapper.instance().handleFormSubmit(formValues, {})

      // then
      expect(wrapper.state('isLoadingMode')).toBe(true)
      expect(createVenueProvider).toHaveBeenCalledWith(expect.any(Function), expect.any(Function), {
        providerId: 'CC',
        venueId: 'AA',
        venueIdAtOfferProvider: 'token',
      })
    })
  })

  describe('handleSuccess', () => {
    it('should update current url when action was handled successfully', () => {
      // given
      const wrapper = shallow(<TiteliveProviderForm {...props} />)

      // when
      wrapper.instance().handleSuccess()

      // then
      expect(history.push).toHaveBeenCalledWith('/structures/CC/lieux/AA')
    })
  })

  describe('handleFail', () => {
    it('should display a notification with the proper informations', () => {
      // given
      const wrapper = shallow(<TiteliveProviderForm {...props} />)
      const action = {
        payload: {
          errors: [
            {
              error: 'fake error',
            },
          ],
        },
      }
      const form = {
        batch: jest.fn(),
      }
      // when
      wrapper.instance().handleFail(form)({}, action)
      // then
      expect(notify).toHaveBeenCalledWith([{ error: 'fake error' }])
    })
  })
})
