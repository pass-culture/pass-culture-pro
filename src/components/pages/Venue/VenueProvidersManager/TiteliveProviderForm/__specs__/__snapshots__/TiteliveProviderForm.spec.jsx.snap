// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | pages | Venue | VenueProvidersManager | form | TiteliveProviderForm should match the snapshot 1`] = `ShallowWrapper {}`;
